### PR TITLE
[Fix-17040][sql] Preserve PL/pgSQL DO $$ blocks in PostgreSQL SQL splitting

### PR DESCRIPTION
**Pull request by DhiyaneshwaranR**
## Purpose of the pull request

Fixes issue [#17040](https://github.com/apache/dolphinscheduler/issues/17040)

This PR fixes incorrect splitting of PL/pgSQL `DO $$...$$;` blocks in PostgreSQL SQL tasks.

### Root cause

The regression was introduced in [#15367](https://github.com/apache/dolphinscheduler/pull/15367), which enabled Druid-based SQL splitting.  
However, Druid's default SQL parser splits on semicolons — including those inside dollar-quoted `DO $$...$$` blocks — which breaks valid PL/pgSQL scripts by triggering errors like `Unterminated dollar quote`.

---

## Brief change log

- Override `splitAndRemoveComment()` in `PostgreSQLDataSourceProcessor`
- Implement custom dollar-quote-aware SQL splitter
- Preserve semicolons inside PL/pgSQL `DO $$ ... $$` blocks
- Leave standard SQL splitting behaviour unchanged
- Add unit tests to verify proper behaviour for multi-statement scripts

---

## Verify this pull request

This change added tests and can be verified as follows:

- ✅ Manually tested using DolphinScheduler UI SQL task with:
  - Single `DO $$ ... $$;` block
  - `DO` block followed by `INSERT` and `UPDATE`
- ✅ Added unit tests in `PostgreSQLDataSourceProcessorTest`:
  - DO block only
  - DO block + INSERT
  - Dollar-tagged blocks (`$func$`)
  - Standard SQL with multiple statements

---

## Documentation

This pull request does **not** require documentation update — it only fixes internal SQL splitting logic for PostgreSQL.

---

## Incompatible Changes

This PR does **not** introduce any incompatible changes.

### CODE REVIEW

**Comment by DhiyaneshwaranR** [comment]
Hi @chaz6chez @SbloodyS 👋

This PR addresses issue [#17040](https://github.com/apache/dolphinscheduler/issues/17040) by correctly handling PostgreSQL `DO $$...$$;` blocks that were being split incorrectly by Druid.

- It introduces a dollar-quote-aware SQL splitter in `PostgreSQLDataSourceProcessor`
- Semicolons inside PL/pgSQL blocks are now preserved properly
- Regular multi-statement SQL still works as expected
- Unit tests and manual verification via the UI have been added ✅

Please have a look when you get time. Would love your feedback 🙌

**Comment by DhiyaneshwaranR** [comment]
Hi @SbloodyS, I’ve pushed the latest changes with spotless and license fixes — all set now.
Would be great if you could take a quick look when you're free.
Thanks again!

**Comment by SbloodyS** [comment]
> Hi @SbloodyS, I’ve pushed the latest changes with spotless and license fixes — all set now. Would be great if you could take a quick look when you're free. Thanks again!

I'm not quite familiar with postgresql. PTAL @Gallardot @ruanwenjun @caishunfeng

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-postgresql/src/test/java/org/apache/dolphinscheduler/plugin/datasource/postgresql/PostgreSQLDataSourceProcessorTest.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+public class PostgreSQLDataSourceProcessorTest {
+
+    private final PostgreSQLDataSourceProcessor processor = new PostgreSQLDataSourceProcessor();
+
+    @Test
+    public void testSingleDoBlock() {
+        String sql = "DO $$ BEGIN RAISE NOTICE 'Hello'; END $$;";
+        List<String> parts = processor.splitAndRemoveComment(sql);
+        Assertions.assertEquals(1, parts.size());
+        Assertions.assertTrue(parts.get(0).startsWith("DO $$"));
```
Comment (ruanwenjun) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-postgresql/src/test/java/org/apache/dolphinscheduler/plugin/datasource/postgresql/PostgreSQLDataSourceProcessorTest.java]: It might better assert the full content rather than prefix.

**Review by ruanwenjun — COMMENTED** [review]
Thanks for your PR.

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-postgresql/src/test/java/org/apache/dolphinscheduler/plugin/datasource/postgresql/PostgreSQLDataSourceProcessorTest.java (hunk 1 above)
Comment (DhiyaneshwaranR) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-postgresql/src/test/java/org/apache/dolphinscheduler/plugin/datasource/postgresql/PostgreSQLDataSourceProcessorTest.java]: Hope this is fine . I have updated the test

**Comment by DhiyaneshwaranR** [comment]
Hi @ruanwenjun and @SbloodyS — I’ve pushed the latest changes to address the review feedback.
Let me know if anything else needs attention. Hopefully this finalises everything in the PR.
Appreciate your time and support!

File: .idea/vcs.xml [hunk 2]
```diff
@@ -12,5 +12,10 @@
   </component>
   <component name="VcsDirectoryMappings">
     <mapping directory="" vcs="Git" />
+    <mapping directory="$PROJECT_DIR$/.github/actions/auto-assign-action" vcs="Git" />
```
Comment (ruanwenjun) [inline .idea/vcs.xml]: Please rollback this file.

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-postgresql/src/test/java/org/apache/dolphinscheduler/plugin/datasource/postgresql/PostgreSQLDataSourceProcessorTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        String sql = "DO $$ BEGIN RAISE NOTICE 'Start'; END $$;\nINSERT INTO test_table(id) VALUES (1);";
+        List<String> parts = processor.splitAndRemoveComment(sql);
+        Assertions.assertEquals(2, parts.size());
+    }
+
+    @Test
+    public void testStandardSql() {
+        String sql = "INSERT INTO test_table(id) VALUES (1); UPDATE test_table SET id = 2 WHERE id = 1;";
+        List<String> parts = processor.splitAndRemoveComment(sql);
+        Assertions.assertEquals(2, parts.size());
```
Comment (ruanwenjun) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-postgresql/src/test/java/org/apache/dolphinscheduler/plugin/datasource/postgresql/PostgreSQLDataSourceProcessorTest.java:50]:
```suggestion
        Assertions.assertEquals(2, parts.size());
        Assertions.assertEquals("INSERT INTO test_table(id) VALUES (1)", parts.get(0));
        Assertions.assertEquals("UPDATE test_table SET id = 2 WHERE id = 1", parts.get(1));
```

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-postgresql/src/test/java/org/apache/dolphinscheduler/plugin/datasource/postgresql/PostgreSQLDataSourceProcessorTest.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @Test
+    public void testCustomDollarTag() {
+        String sql = "DO $func$ BEGIN RAISE NOTICE 'Tag test'; END $func$;";
+        List<String> parts = processor.splitAndRemoveComment(sql);
+        Assertions.assertEquals(1, parts.size());
+        Assertions.assertTrue(parts.get(0).contains("$func$"));
```
Comment (ruanwenjun) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-postgresql/src/test/java/org/apache/dolphinscheduler/plugin/datasource/postgresql/PostgreSQLDataSourceProcessorTest.java:60]:
```suggestion
        Assertions.assertTrue("DO $func$ BEGIN RAISE NOTICE 'Tag test'; END $func$", parts.get(0));
```

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-postgresql/src/test/java/org/apache/dolphinscheduler/plugin/datasource/postgresql/PostgreSQLDataSourceProcessorTest.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @Test
+    public void testCommentsPreserved() {
+        String sql =
+                "-- comment here\nDO $$ BEGIN NULL; END $$;\n-- trailing comment\nINSERT INTO test_table VALUES (5);";
+        List<String> parts = processor.splitAndRemoveComment(sql);
+        Assertions.assertEquals(2, parts.size());
```
Comment (ruanwenjun) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-postgresql/src/test/java/org/apache/dolphinscheduler/plugin/datasource/postgresql/PostgreSQLDataSourceProcessorTest.java:69]:
```suggestion
        Assertions.assertEquals(2, parts.size());
        Assertions.assertEquals("DO $$ BEGIN NULL; END $$", parts.get(0));
        Assertions.assertEquals("INSERT INTO test_table VALUES (5)", parts.get(1));
```

**Comment by DhiyaneshwaranR** [comment]
Hi @ruanwenjun, I’ve updated the logic based on a suggestion from Copilot and adjusted the test cases as per your feedback. I’ve also added a few more test cases to cover additional edge scenarios.

Hopefully, everything looks good now and this finalises the PR.
Thanks again for your support and helpful feedback!

**Review by ruanwenjun — APPROVED** [review]
LGTM

**Review by SbloodyS — APPROVED** [review]
+1
